tardis/transport/montecarlo/formal_integral_cuda.py

```diff
@@ -12,6 +12,212 @@
 H_CGS = 6.62606957e-27
 
 
+@cuda.jit
+def cuda_vector_integrator(L, I_nu, N, R_max):
+
+    nu_idx = cuda.grid(1)
+    L[nu_idx] = (
+        8 * M_PI * M_PI * trapezoid_integration_cuda(I_nu[nu_idx], R_max / N)
+    )
+
+@cuda.jit
+def cuda_formal_integral_fast(
+    r_inner,
+    r_outer,
+    time_explosion,
+    line_list_nu,
+    iT,
+    inu,
+    inu_size,
+    att_S_ul,
+    Jred_lu,
+    Jblue_lu,
+    tau_sobolev,
+    electron_density,
+    N,
+    pp,
+    exp_tau,
+    I_nu,
+    z,
+    shell_id,
+):
+    """
+    The CUDA version of numba_formal_integral that can run
+    on a NVIDIA GPU.
+
+    Parameters
+    ----------
+    r_inner : array(float64, 1d, C)
+        self.geometry.r_inner
+    r_outer : array(float64, 1d, C)
+        self.geometry.r_outer
+    time_explosion: float64
+        self.geometry.time_explosion
+    line_list_nu : array(float64, 1d, A)
+        self.plasma.line_list_nu
+    iT : np.float64
+    inu : np.float64
+    inu_size : int64
+    att_S_ul : array(float64, 1d, C)
+    Jred_lu : array(float64, 1d, C)
+    Jblue_lu : array(float64, 1d, C)
+    tau_sobolev : array(float64, 2d, C)
+    electron_density : array(float64, 1d, C)
+    N : int64
+    L : array(float64, 1d, C)
+        This is where the results will be stored
+    pp : array(float64, 1d, C)
+    exp_tau : array(float64, 1d, C)
+    I_nu array(floatt64, 2d, C)
+    z : array(float64, 2d, C)
+    shell_id : array(int64, 2d, C)
+    """
+
+    # todo: add all the original todos
+
+    # global read-only values
+    size_line, size_shell = tau_sobolev.shape
+    size_tau = size_line * size_shell
+    R_ph = r_inner[0]  # make sure these are cgs
+    R_max = r_outer[size_shell - 1]
+
+    nu_idx, p_idx = cuda.grid(2) # 2D Cuda Grid, nu x p
+    p_idx += 1 # Because the iteration starts at 1
+    # Check to see if CUDA is out of bounds
+    if nu_idx >= inu_size:
+        return
+
+    if p_idx >= N:
+        return
+
+    # These all have their own version for each thread to avoid race conditions
+    I_nu_thread = I_nu[nu_idx]
+    z_thread = z[nu_idx]
+    shell_id_thread = shell_id[nu_idx]
+
+    offset = 0
+    size_z = 0
+    idx_nu_start = 0
+    direction = 0
+    first = 0
+    i = 0
+    p = 0.0
+    nu_start = 0.0
+    nu_end = 0.0
+    nu = 0.0
+    zstart = 0.0
+    zend = 0.0
+    escat_contrib = 0.0
+    escat_op = 0.0
+    Jkkp = 0.0
+    pexp_tau = 0
+    patt_S_ul = 0
+    pJred_lu = 0
+    pJblue_lu = 0
+    pline = 0
+
+    nu = inu[nu_idx]
+
+    # now loop over discrete values along line
```

Comment:
Along which line? I don't see a loop here

Reply:
Yeah, the loop is in the GPU threads, but I'll remove the comment